src/main/scala/is/hail/io/vcf/LoadVCF.scala

```diff
@@ -188,6 +277,245 @@ final class VCFLine(line: String) {
+      mk = true
+      pos += 1
+    } else
+      k = parseIntInCall()
+
+    if (!endFormatField()) {
+      if (line(pos) == '/' || line(pos) == '|')
+        fatal("ploidy > 2 not supported")
+      else
+        fatal("parse error in call")
+    }
+
+    // treat partially missing like missing
+    if (mj || mk)
+      rvb.setMissing()
+    else {
+      rvb.addInt(Genotype.gtIndexWithSwap(j, k))
+    }
+  }
+
+  def parseFormatInt(): Int = {
+    if (endFormatField())
+      fatal("empty integer")
+    var v = VCFLine.numericValue(line(pos))
+    pos += 1
+    while (!endFormatField()) {
+      v = v * 10 + VCFLine.numericValue(line(pos))
+      pos += 1
+    }
+    v
+  }
+
+  def parseAddFormatInt(rvb: RegionValueBuilder) {
+    if (formatFieldMissing()) {
+      rvb.setMissing()
+      pos += 1
+    } else
+      rvb.addInt(parseFormatInt())
+  }
+
+  def parseFormatString(): String = {
+    val start = pos
+    while (!endFormatField())
+      pos += 1
+    val end = pos
+    line.substring(start, end)
+  }
+
+  def parseAddFormatString(rvb: RegionValueBuilder) {
+    if (formatFieldMissing()) {
+      rvb.setMissing()
+      pos += 1
+    } else
+      rvb.addString(parseFormatString())
+  }
+
+  def parseFormatDouble(): Double = {
+    val s = parseFormatString()
+    s.toDouble
+  }
+
+  def parseAddFormatDouble(rvb: RegionValueBuilder) {
+    if (formatFieldMissing()) {
+      rvb.setMissing()
+      pos += 1
+    } else
+      rvb.addDouble(parseFormatDouble())
+  }
+
+  def parseIntInFormatArray(rvb: RegionValueBuilder): Int = {
+    if (endFormatArrayField())
+      fatal("empty integer")
+    var v = VCFLine.numericValue(line(pos))
+    pos += 1
+    while (!endFormatArrayField()) {
+      v = v * 10 + VCFLine.numericValue(line(pos))
+      pos += 1
+    }
+    v
+  }
+
+  def parseAddFormatArrayInt(rvb: RegionValueBuilder) {
+    if (formatFieldMissing()) {
+      rvb.setMissing()
+      pos += 1
+    } else {
+      assert(abi.length == 0)
+
+      abi += parseIntInFormatArray(rvb)
+      while (!endFormatField()) {
+        pos += 1 // comma
+        abi += parseIntInFormatArray(rvb)
+      }
+
+      rvb.startArray(abi.length)
+      var i = 0
+      while (i < abi.length) {
+        rvb.addInt(abi(i))
+        i += 1
+      }
+      rvb.endArray()
+
+      abi.clear()
+    }
+  }
+}
+
+object FormatParser {
+  def apply(gType: TStruct, format: String): FormatParser = {
+    val formatFields = format.split(":")
+    val formatFieldsSet = formatFields.toSet
+    new FormatParser(
+      gType,
+      formatFields.map(gType.fieldIdx),
+      (0 until gType.size)
+        .filter(i => !formatFieldsSet.contains(gType.fields(i).name))
+        .toArray)
+  }
+}
+
+class FormatParser(
+  gType: TStruct,
+  formatFieldGIndex: Array[Int],
+  missingGIndices: Array[Int]) {
+
+  def parseAddField(l: VCFLine, rvb: RegionValueBuilder, i: Int) {
+    val j = formatFieldGIndex(i)
+    rvb.setFieldIndex(j)
+    gType.fieldType(j) match {
```

Comment:
this part is probably slow, should be fixed in a future PR

Reply:
Agreed.  Staging!  In C!